In [2]:
import numpy as np
import pandas as pd
import geopandas as gpd
import requests
import pygeos
import osmnx as ox
import shapely as shp
from ast import literal_eval
from pyproj import Geod
import matplotlib.pyplot as plt
import folium

Let’s use a newer dataset https://health.data.ny.gov/Health/Institutional-Provider-Network-Data-2021-Quarter-1/tfz4-6xw3 

I think each record represents a physician accepting different healthcare insurance. 

1. Find records with the following criteria.
    - ‘Designated Service’: 250 (Emergency Medication) or 823 (Urgent Care) 
    - or ‘Site Name’ has Emergency or Urgent (regardless of Designated Service)
2. Count the number of providers per facility using ‘NPI’ and ‘Site Name’
    - Exclude ‘NPI’ values 9999999999 & 8888888888 or some other invalid values
    - Exclude ‘Site Name’ is NaN. 
3. Final products
    - Doctor counts per facility: Hospital_ID, Address, Lat, Long, ProviderCount
    - List of Doctors (NPI) associated with each facility: Hospital_ID, Address, NPI


In [3]:
# Census tracts in NYC 
tracts = gpd.read_file('./data/new_york_census_tract.geojson')
tracts['GEOID_County'] = tracts.apply(lambda x:x['GEOID'][0:5], axis=1)
tracts.head(3)

GEOID  area_sqkm                                           geometry  \
0  36081003400   0.213985  POLYGON ((311370.732 57391.499, 311384.757 573...   
1  36081003900   0.146969  POLYGON ((304857.478 66110.315, 304948.710 663...   
2  36081003600   0.180006  POLYGON ((311522.128 57288.347, 311596.477 573...   

  GEOID_County  
0        36081  
1        36081  
2        36081

In [4]:
# Provider data 
providers = pd.read_csv("data/Institutional_Provider_Network_Data__2021_Quarter_1.csv")
providers = gpd.GeoDataFrame(providers, geometry=gpd.points_from_xy(providers['Longitude'], providers['Latitude'], crs="EPSG:4326"))
providers['Designated Service'] = providers['Designated Service'].apply(lambda x: '{0:0>3}'.format(x))
providers = providers.to_crs(tracts.crs)

providers.head(3)

/var/folders/dv/bcrkszqj0634wz_17m9xw0qr0000gn/T/ipykernel_26482/779929108.py:2: DtypeWarning: Columns (14,15,16,80) have mixed types. Specify dtype option on import or set low_memory=False.
  providers = pd.read_csv("data/Institutional_Provider_Network_Data__2021_Quarter_1.csv")


Site Name Room/Suite  \
0                                       02 SOLUTIONS        NaN   
1                                       02 SOLUTIONS        NaN   
2  1 BETHESDA DRIVE OPERATING COMPANY LLC DB ELDE...        NaN   

         Address     Town/City  CountyFIPS County Name State  Zip Code  \
0  1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   12065.0   
1  1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   12065.0   
2  1 BETHESDA DR       HORNELL     36101.0     Steuben    NY   14843.0   

    Latitude  Longitude  ...  Servicing County 3  Servicing County 4  \
0  42.825080 -73.733701  ...                  35                  39   
1  42.825080 -73.733701  ...                 113                 115   
2  42.346198 -77.665526  ...                 999                 999   

  Servicing County 5  OASAS Program Reporting Unit Number  \
0                 57                                  NaN   
1                999                                  NaN   
2                999                                  NaN   

  OASAS Provider Number OMH ID                                Plan Name  \
0                   NaN    NaN                 Nascentia Health Options   
1                   NaN    NaN                 Nascentia Health Options   
2                   NaN    NaN  New York Quality Healthcare Corporation   

   Provider Network Data System Identification Number    Quarter/Year  \
0                                                 91   Quarter 1 2021   
1                                                 91   Quarter 1 2021   
2                                                 44   Quarter 1 2021   

                        geometry  
0  POINT (321788.653 295329.915)  
1  POINT (321788.653 295329.915)  
2   POINT (-2053.454 248396.682)  

[3 rows x 85 columns]

In [5]:
providers

Site Name  Room/Suite  \
0                                            02 SOLUTIONS         NaN   
1                                            02 SOLUTIONS         NaN   
2       1 BETHESDA DRIVE OPERATING COMPANY LLC DB ELDE...         NaN   
3       1 BETHESDA DRIVE OPERATING COMPANY LLC DB ELDE...         NaN   
4       1 BETHESDA DRIVE OPERATING COMPANY LLC DB ELDE...         NaN   
...                                                   ...         ...   
570807                                ZWH MEDICAL CARE PC         NaN   
570808                     ZYCARE MEDICAL ASSOCIATES PLLC  Suite CS 8   
570809        ZYGMUNT DOROZYNSKI, DDS FAMILY DENTISTRY PC         NaN   
570810        ZYGMUNT DOROZYNSKI, DDS FAMILY DENTISTRY PC         NaN   
570811                                   ZYNITECH MEDICAL         NaN   

                     Address     Town/City  CountyFIPS County Name State  \
0              1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   
1              1406B ROUTE 9  CLIFTON PARK     36091.0    Saratoga    NY   
2              1 BETHESDA DR       HORNELL     36101.0     Steuben    NY   
3              1 BETHESDA DR       HORNELL     36101.0     Steuben    NY   
4              1 BETHESDA DR       HORNELL     36101.0     Steuben    NY   
...                      ...           ...         ...         ...   ...   
570807         6417 BROADWAY      WOODSIDE     36081.0      Queens    NY   
570808         13338 41ST RD      FLUSHING     36081.0      Queens    NY   
570809        130 N FRONT ST      KINGSTON     36111.0      Ulster    NY   
570810        130 N FRONT ST      KINGSTON     36111.0      Ulster    NY   
570811  347 N POTTSTOWN PIKE         EXTON     42029.0     Chester    PA   

        Zip Code   Latitude  Longitude  ...  Servicing County 3  \
0        12065.0  42.825080 -73.733701  ...                  35   
1        12065.0  42.825080 -73.733701  ...                 113   
2        14843.0  42.346198 -77.665526  ...                 999   
3        14843.0  42.346198 -77.665526  ...                 999   
4        14843.0  42.346198 -77.665526  ...                 999   
...          ...        ...        ...  ...                 ...   
570807   11377.0  40.749624 -73.897965  ...                 999   
570808   11355.0  40.756774 -73.830378  ...                 999   
570809   12401.0  41.934896 -74.024026  ...                 999   
570810   12401.0  41.934896 -74.024026  ...                 999   
570811   19341.0  40.039158 -75.632274  ...                 999   

        Servicing County 4 Servicing County 5  \
0                       39                 57   
1                      115                999   
2                      999                999   
3                      999                999   
4                      999                999   
...                    ...                ...   
570807                 999                999   
570808                 999                999   
570809                 999                999   
570810                 999                999   
570811                 999                999   

        OASAS Program Reporting Unit Number OASAS Provider Number OMH ID  \
0                                       NaN                   NaN    NaN   
1                                       NaN                   NaN    NaN   
2                                       NaN                   NaN    NaN   
3                                       NaN                   NaN    NaN   
4                                       NaN                   NaN    NaN   
...                                     ...                   ...    ...   
570807                                  NaN                   NaN    NaN   
570808                                  NaN                   NaN    NaN   
570809                                  NaN                   NaN    NaN   
570810                                  NaN                   NaN    NaN   
570811                      

In [6]:
# Provider wiithin NYC
tract_geom_union = pygeos.from_shapely(tracts['geometry'].unary_union)

provider_geom = pygeos.from_shapely(providers['geometry'].values)    
provider_geom_tree = pygeos.STRtree(provider_geom, leafsize=50)

query_intersect = provider_geom_tree.query(tract_geom_union, predicate='contains')
providers_nyc = providers.loc[query_intersect]
providers_nyc.head(3)

Site Name Room/Suite            Address  \
1029    210 MJD PRECISION URGENT CARE        NaN  1737 SEAGIRT BLVD   
463720    SOUTH ISLAND MED ASSOCIATES        NaN  1735 SEAGIRT BLVD   
414968    RENAL DIVISION OF MT. SINAI        NaN  1731 SEAGIRT BLVD   

           Town/City  CountyFIPS County Name State  Zip Code   Latitude  \
1029    FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
463720  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
414968  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   

        Longitude  ...  Servicing County 3  Servicing County 4  \
1029   -73.751462  ...                 999                 999   
463720 -73.751460  ...                 999                 999   
414968 -73.751456  ...                 999                 999   

       Servicing County 5  OASAS Program Reporting Unit Number  \
1029                  999                                  NaN   
463720                999                                  NaN   
414968                999                                  NaN   

       OASAS Provider Number OMH ID                             Plan Name  \
1029                     NaN    NaN  Consolidated Health Plans: MagnaCare   
463720                   NaN    NaN    UnitedHealthcare of New York, Inc.   
414968                   NaN    NaN                 Oscar Insurance Corp.   

        Provider Network Data System Identification Number    Quarter/Year  \
1029                                                  169   Quarter 1 2021   
463720                                                148   Quarter 1 2021   
414968                                                 75   Quarter 1 2021   

                            geometry  
1029    POINT (321038.397 47583.142)  
463720  POINT (321038.567 47583.143)  
414968  POINT (321038.905 47583.144)  

[3 rows x 85 columns]

In [7]:
providers_nyc

Site Name Room/Suite               Address  \
1029     210 MJD PRECISION URGENT CARE        NaN     1737 SEAGIRT BLVD   
463720     SOUTH ISLAND MED ASSOCIATES        NaN     1735 SEAGIRT BLVD   
414968     RENAL DIVISION OF MT. SINAI        NaN     1731 SEAGIRT BLVD   
329494           NEW HEALTH ASSOCIATES        NaN     1731 SEAGIRT BLVD   
319789  MOUNT SINAI HOSPITAL-PROF FEES        NaN     1731 SEAGIRT BLVD   
...                                ...        ...                   ...   
23211     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
23212     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
23216     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
82571       CATHOLIC GUARDIAN SERVICES        NaN        743 E 243RD ST   
536977       WAKEFIELD PEDIATRICS PLLC        NaN  4813 WHITE PLAINS RD   

           Town/City  CountyFIPS County Name State  Zip Code   Latitude  \
1029    FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
463720  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
414968  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
329494  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
319789  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
...              ...         ...         ...   ...       ...        ...   
23211          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
23212          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
23216          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
82571          BRONX     36005.0       Bronx    NY   10470.0  40.906027   
536977         BRONX     36005.0       Bronx    NY   10470.0  40.906253   

        Longitude  ...  Servicing County 3  Servicing County 4  \
1029   -73.751462  ...                 999                 999   
463720 -73.751460  ...                 999                 999   
414968 -73.751456  ...                 999                 999   
329494 -73.751456  ...                 999                 999   
319789 -73.751456  ...                 999                 999   
...           ...  ...                 ...                 ...   
23211  -73.849340  ...                  81                  61   
23212  -73.849340  ...                  81                  61   
23216  -73.849340  ...                  81                  61   
82571  -73.847213  ...                 999                 999   
536977 -73.849314  ...                 999                 999   

       Servicing County 5  OASAS Program Reporting Unit Number  \
1029                  999                                  NaN   
463720                999                                  NaN   
414968                999                                  NaN   
329494                999                                  NaN   
319789                999                                  NaN   
...                   ...                                  ...   
23211                  85                                  NaN   
23212                  85                                  NaN   
23216                  85                                  NaN   
82571                 999                                  NaN   
536977                999                                  NaN   

       OASAS Provider Number OMH ID  \
1029                     NaN    NaN   
463720                   NaN    NaN   
414968                   NaN    NaN   
329494                   NaN    NaN   
319789                   NaN    NaN   
...                      ...    ...   
23211                    NaN    NaN   
23212                    NaN    NaN   
23216                    NaN    NaN   
82571                    NaN    NaN   
536977                   NaN    NaN   

                                                Plan Name  \
1029                 Consolidated Health Plans: MagnaCare   
463720                 UnitedHealthcare of New Yor

In [5]:
providers_nyc['NPI'].value_counts()

9999999999    6807
8888888888    4554
1821036807    1985
1366430563    1716
1831269315    1596
              ... 
1780121228       1
1972042901       1
1669710117       1
1073721130       1
1558532796       1
Name: NPI, Length: 11436, dtype: int64

In [6]:
providers_nyc.loc[providers_nyc['NPI'] == 1821036807]

Site Name  \
273345               LENOX HILL RADIO MED IMAGING ASSO PC   
274334  LENOX HILL RADIOLOGY & MEDICAL IMAGING ASSOCIATES   
273563                               LENOX HILL RADIOLOGY   
273564                               LENOX HILL RADIOLOGY   
275793             LENOX HILL RADIOLOGY MEDICAL IMAGING A   
...                                                   ...   
274206    LENOX HILL RADIOLOGY - WASHINGTON HEIGHTS NORTH   
274725  LENOX HILL RADIOLOGY & MEDICAL IMAGING ASSOCIA...   
274726  LENOX HILL RADIOLOGY & MEDICAL IMAGING ASSOCIA...   
278457                    LHR OF WASHINGTON HEIGHTS NORTH   
274283  LENOX HILL RADIOLOGY & MEDICAL IMAGING ASSOCIATES   

                  Room/Suite           Address Town/City  CountyFIPS  \
273345                   NaN  2832 LINDEN BLVD  BROOKLYN     36047.0   
274334                   NaN  2832 LINDEN BLVD  BROOKLYN     36047.0   
273563                   NaN  2832 LINDEN BLVD  BROOKLYN     36047.0   
273564                   NaN  2832 LINDEN BLVD  BROOKLYN     36047.0   
275793        PHP-LINDENWOOD  2832 LINDEN BLVD  BROOKLYN     36047.0   
...                      ...               ...       ...         ...   
274206                   NaN     4334 BROADWAY  NEW YORK     36061.0   
274725  LHR of Washington He     4334 BROADWAY  NEW YORK     36061.0   
274726  LHR of Washington He     4334 BROADWAY  NEW YORK     36061.0   
278457                 STE 1     4334 BROADWAY  NEW YORK     36061.0   
274283                   NaN   4234 BRONX BLVD     BRONX     36005.0   

       County Name State  Zip Code   Latitude  Longitude  ...  \
273345       Kings    NY   11208.0  40.669604 -73.859551  ...   
274334       Kings    NY   11208.0  40.669604 -73.859551  ...   
273563       Kings    NY   11208.0  40.669604 -73.859551  ...   
273564       Kings    NY   11208.0  40.669604 -73.859551  ...   
275793       Kings    NY   11208.0  40.669604 -73.859551  ...   
...            ...   ...       ...        ...        ...  ...   
274206    New York    NY   10033.0  40.852974 -73.934110  ...   
274725    New York    NY   10033.0  40.852974 -73.934110  ...   
274726    New York    NY   10033.0  40.852974 -73.934110  ...   
278457    New York    NY   10033.0  40.852974 -73.934110  ...   
274283       Bronx    NY   10466.0  40.895546 -73.860387  ...   

        Servicing County 3  Servicing County 4 Servicing County 5  \
273345                 999                 999                999   
274334                 999                 999                999   
273563                 999                 999                999   
273564                 999                 999                999   
275793                 999                 999                999   
...                    ...                 ...                ...   
274206                 999                 999                999   
274725                 999                 999                999   
274726                 999                 999                999   
278457                 999                 999                999   
274283                 999                 999                999   

        OASAS Program Reporting Unit Number OASAS Provider Number OMH ID  \
273345                                  NaN                   NaN    NaN   
274334                                  NaN                   NaN    NaN   
273563                                  NaN                   NaN    NaN   
273564                                  NaN                   NaN    NaN   
275793                                  NaN                   NaN    NaN   
...                                     ...                   ...    ...   
274206                                  NaN                   NaN    NaN   
274725                                  NaN                   NaN    NaN   
274726                                  NaN                   NaN    NaN   
278457                                  NaN                   NaN    NaN   
2742

In [7]:
providers_nyc['Designated Service'].value_counts()

760    84150
914    36569
375    13320
307     7725
911     6958
       ...  
835        1
205        1
084        1
915        1
981        1
Name: Designated Service, Length: 259, dtype: int64

In [8]:
# Exclude invalid values
providers_nyc = providers_nyc.loc[~providers_nyc['Site Name'].isna()]
providers_nyc

Site Name Room/Suite               Address  \
1029     210 MJD PRECISION URGENT CARE        NaN     1737 SEAGIRT BLVD   
463720     SOUTH ISLAND MED ASSOCIATES        NaN     1735 SEAGIRT BLVD   
414968     RENAL DIVISION OF MT. SINAI        NaN     1731 SEAGIRT BLVD   
329494           NEW HEALTH ASSOCIATES        NaN     1731 SEAGIRT BLVD   
319789  MOUNT SINAI HOSPITAL-PROF FEES        NaN     1731 SEAGIRT BLVD   
...                                ...        ...                   ...   
23211     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
23212     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
23216     ALPINE HOME HEALTH CARE, LLC        NaN  4770 WHITE PLAINS RD   
82571       CATHOLIC GUARDIAN SERVICES        NaN        743 E 243RD ST   
536977       WAKEFIELD PEDIATRICS PLLC        NaN  4813 WHITE PLAINS RD   

           Town/City  CountyFIPS County Name State  Zip Code   Latitude  \
1029    FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
463720  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
414968  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
329494  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
319789  FAR ROCKAWAY     36081.0      Queens    NY   11691.0  40.594899   
...              ...         ...         ...   ...       ...        ...   
23211          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
23212          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
23216          BRONX     36005.0       Bronx    NY   10470.0  40.905224   
82571          BRONX     36005.0       Bronx    NY   10470.0  40.906027   
536977         BRONX     36005.0       Bronx    NY   10470.0  40.906253   

        Longitude  ...  Servicing County 3  Servicing County 4  \
1029   -73.751462  ...                 999                 999   
463720 -73.751460  ...                 999                 999   
414968 -73.751456  ...                 999                 999   
329494 -73.751456  ...                 999                 999   
319789 -73.751456  ...                 999                 999   
...           ...  ...                 ...                 ...   
23211  -73.849340  ...                  81                  61   
23212  -73.849340  ...                  81                  61   
23216  -73.849340  ...                  81                  61   
82571  -73.847213  ...                 999                 999   
536977 -73.849314  ...                 999                 999   

       Servicing County 5  OASAS Program Reporting Unit Number  \
1029                  999                                  NaN   
463720                999                                  NaN   
414968                999                                  NaN   
329494                999                                  NaN   
319789                999                                  NaN   
...                   ...                                  ...   
23211                  85                                  NaN   
23212                  85                                  NaN   
23216                  85                                  NaN   
82571                 999                                  NaN   
536977                999                                  NaN   

       OASAS Provider Number OMH ID  \
1029                     NaN    NaN   
463720                   NaN    NaN   
414968                   NaN    NaN   
329494                   NaN    NaN   
319789                   NaN    NaN   
...                      ...    ...   
23211                    NaN    NaN   
23212                    NaN    NaN   
23216                    NaN    NaN   
82571                    NaN    NaN   
536977                   NaN    NaN   

                                                Plan Name  \
1029                 Consolidated Health Plans: MagnaCare   
463720                 UnitedHealthcare of New Yor

In [9]:
# You probably can do something like this
unique_df = pd.DataFrame(providers_nyc[['Site Name', 'Address', 'NPI']].value_counts().reset_index())
unique_df

Site Name              Address  \
0      THE DORAL INVESTORS GROUP LLC DBA HOUSE CALLS ...       2440 FULTON ST   
1                                      PEOPLE CARE, INC.        116 W 32ND ST   
2                               ELMHURST HOSPITAL CENTER        7901 BROADWAY   
3                                             THE BRIDGE       248 W 108TH ST   
4                          NEW HORIZON COUNSELING CENTER    718 BEACH 20TH ST   
...                                                  ...                  ...   
53110               KUMP & SAYEGH FAMILY MEDICAL SVCS PC       530 W 236TH ST   
53111                             KUMAR PHYSICIAN OFFICE  25012B HILLSIDE AVE   
53112                           KUMAR MEDICAL CARE, P.C.        2310 30TH AVE   
53113                           KUMAR MEDICAL CARE, P.C.    12415 JAMAICA AVE   
53114                     ZYCARE MEDICAL ASSOCIATES PLLC        13338 41ST RD   

              NPI    0  
0      1639473580  500  
1      1841302445  373  
2      1063426377  229  
3      1831269315  223  
4      1366430563  203  
...           ...  ...  
53110  1336316348    1  
53111  1447628342    1  
53112  1548601024    1  
53113  1548601024    1  
53114  1639134190    1  

[53115 rows x 4 columns]

## Notes



benefits over health facility genearl information:

- more complete (e.g., has urgent care)
- have more in-depth segmentation (instead of clinic, have type of clinic and what it actually does, which can inform whether a facility supports walk-ins), (desser + service types)
    - e.g., hospital extension clinics in hfgi include a methadone center, a dance injury center, and a dentistry center while in pnds, they are segmented into more identifying categories
- includes the number of providers (noprov)

cons versus health facility general information

- more complex (has a bunch of random, irrelevant stuff)
- indexed by health facilities & plans based on which plans contract with what facilities -- not directly comparable to healthcare accessibility without some tinkering

documentation: https://www.health.ny.gov/health_care/managed_care/docs/dictionary.pdf

what columns mean: https://www.opendatanetwork.com/dataset/health.data.ny.gov/eary-k6ia

To use this data, we need to assume the following:

1. The institutions that are involved in group medical insurance plans are representative of all relevant institutions.
    - this is somewhat reasonable since these institutions would be the most accessible to the average person
2. The assigned designated service is representative of what a facility actually does and can be used to determine whether a facility supports walk-ins or not
    - (can't guarantee that walk-ins are supported but can make generalizations like that clinics generally allow walk-ins)
3. Some subset of the data can be validly excluded, as it is not relevant to medical accessibility
    - should ideally exclude all medically tangential businesses while including relatively relevant ones (like urgent care)

Potential service codes to look for (that might support walk-ins)

- 005 state operated clinic
- 011 general hospital
- 034 state operated facility
- 150 pediatrics
- 252 primary care initiative in underserved areas
- 760 pharmacy?
- 776 general mpractice only - nospec?
- 777 all physician?
- 823 urgent care
- 899 hospital inpatient?
- 901 emergency room
- 914 general medicine - clinic speciality ?
- 978 preferred primary care clinic?
- clinic specialty things (have basically one for every field)